### ISSUE TITLE
Radix Git client - LFS (large file storage) support

### ISSUE THREAD

**Issue by EqTore** [equinor/radix-platform#246]
Hi, one of my apps (ctdviewer) comes with a data file stored in Git LFS, which is required for the app to run properly. Radix currently doesn't fetch it as part of the clone.
According to Radix support, the Radix Git client does not have support for LFS enabled.

Is it possible to have LFS support enabled in the Git client ?

Thanks,
-Tore

**Comment by dbelyaev** [comment]
+1
LFS is now a de-facto for storing binary files in git repos.

**Comment by emirgens** [comment]
To help us evaluate and clarify the requirement, we ask if you can provide us with more information. Samples of use-cases would be great.

We have now registered this in our backlog, for further evaluation.
https://equinor.atlassian.net/browse/RA-957

**Comment by EqTore** [comment]
Hi,
Thanks for following this up.

GitHub enforces a 100 Mb file size limit when pushing to a repository.
Hence any code base containing binary files larger than 100 Mb needs to use LFS if to be hosted on GitHub.

Since, as far as I know, Radix requires projects to be deployed via GitHub, fetching files from LFS should be supported. It's a quick fix to enable this is in any Git client.

**Comment by trondvs** [comment]
+1
 
Git LFS is also an important feature desired by the Compound Earth Simulator (CES) project. The CES application is written in Matlab and Java and uses the Matlab Compiler SDK to create standalone components that can be deployed without requiring Matlab licensing.

Both the output from the Matlab Compiler SDK and the Matlab runtime libraries it depends on are larger than the 100 Mb limit on github. The repository github.com/equinor/CES therefore uses Git LFS for several files that is required when building the docker image for the app.

Application name in Radix: cesapiserver
Cluster: playground

Best regard,
Trond Vidar Stensby
Kalkulo AS

**Comment by keaaa** [comment]
We've started evaluating if its possible to build using `git LFS` in Radix, within an expected latency to the build operation - could you say something about file size?

**Comment by trondvs** [comment]
The files that presently require Git LFS on github.com/equinor/CES are:

1426 MB, MCR_R2017b_glnxa64_installer.zip, the Matlab Compiler Runtime from MathWorks
472 MB, mkl-libs.tgz, Intel Math Kernel Library
323 MB, cesapiserverMCR.tgz, compiled version of the Compound Earth Simulator (CES), an Equinor product

The first two of these files will be removed from github and replaced by a docker image from a private image hub (docker.pkg.github.com/equinor/ces). These files were initially placed on github since I thought Radix supported Git LFS, but not private image hubs. Now that image hubs are supported it makes sense to have a base docker image that contains Matlab and Intel libraries installed, since this docker image only will change when switching to a newer version of Matlab or Intel MKL.

The third file cesapiserverMCR.tgz will be recreated by our Jenkins build system when pull requests are approved. Creating it depends on the Matlab Compiler SDK, which is a licensed product from MathWorks. Due to the licensing requirements the compilation can not be done on Omnia/Radix and a tar files with the Jenkins artifacts are pushed to github instead.

As a workaround while Git LFS support is unavailable, we will build an additional docker image that also contains cesapiserverMCR.tgz. This image will also be placed in the private image hub. The image builds done in Radix will then be rather trivial with a Dockerfile that simply depends on a image in i private hub.

Audun Solemdal (auls@equinor.com) and Stig Ofstad (stoo@equinor.com) are helping us getting this running on Radix. Stig has created the following illustration of the process that we will be using while Git LFS is unavailable. The "empty" push will be replaced by pushing cesapiserverMCR.tgz when/if Git LFS becomes available.

[cid:9602d611-2fe0-4afe-a70c-6b01099cfca8]

Best regards,

Trond Vidar Stensby
Bluware/Kalkulo


________________________________

**Comment by EqTore** [comment]
Hi Kjell Erik,
Thanks again for addressing this issue.

The repo in question when creating the issue contains one (1) binary file; size is 1.548 Gb.

If this file size is problematic, I will now probably be able to deploy a pre-built image instead of building the image in Radix. Thanks a lot for supporting this.

**Comment by thezultimate** [comment]
A question related to the usage of LFS: Are the big files extracted and processed in the Dockerfile (during build time), or are they only being copied as-is to the container?

**Comment by trondvs** [comment]
In the Compound Earth Simulator (CES) case the big files are extracted when processing the Dockerfile, typically using something like: ADD cesapiserverMCR.tgz /ces

The archive cesapiserverMCR.tgz contains one very large file (approx 300Mb) and various small/medium sized files. This large file is a single jar that is produced by the Matlab Compiler SDK.

**Comment by EqTore** [comment]
Hi (and sorry for the delay),

In my case the big file is a tar.gz which extracts to several data files during build time:
ADD ./static/data/CTDO.tar.gz /CTDViewer/static/data

**Comment by thezultimate** [comment]
Hi everyone,

Thank you for your inputs and information.

We have brainstormed for supporting LFS in Radix and conducted several PoC that actually worked. It is possible to implement this LFS support out-of-the-box in Radix, however, the performance was poor, and this may affect the entire cluster. From this standpoint, we decided to not support LFS inside Radix. We are in the process of implementing deploy-only pipeline in Radix, where users can build Docker images outside of Radix, and deploy them to Radix. We suggest to use this approach when using LFS, and we can help with setting CI outside of Radix if needed.

One question: how critical is this LFS support in Radix for you? Is it a deal-breaker?

If your applications _must_ use LFS in Radix, we have another solution that is a bit hack-ish (thus, we don't recommend), but was tested to work.
